design.md

```diff
@@ -0,0 +1,112 @@
+
+### Scoping Requests
+In order to scope requests to each client we will create a user for each client, and that will be authenticated against the user in the certificate sent from the client.
+
+#### Future Improvements
+* Use a certificate provider for the CA instead of a locally generating one
+* Utilize client certificate details to implement more fine grained access
+* Implement a dedicated secret management solution for private keys, such as Vault or AWS Secret Manager (or any other secure key store)
+* Automate certificate management going forward to remove human error and allow for easier revocation and rotation
+* Add monitoring for certificate expiration and authentication failures
+
+### Library
+The library will primarily utilize the builtin os package in Go to interact with the Linux operating system and accomplish the actual job management functions.
+
+The library will primarily utilize the os/exec builtin, with some use of the os/user and base level os builtin packages.
+
+### Output streaming
+In order to reduce overhead each command will pipe output to a file named with the pid of the process. We will utilize server side streaming from gRPC to serve this data continually to the client until they close the connection or the process finishes.
+
+#### Future Improvements
+* Store output somewhere other than locally on the server, like S3
+* Look into using an external lib that enables tailing from an external source
+* Utilize job management to allow suspending and resuming jobs instead of only being able to start and terminate them
+
+## CLI
+The CLI will utilize the [cobra](https://github.com/spf13/cobra) CLI library.
+
+### Run a process
+The basic structure of the run command will be `run <binary_name> [<args>]`
+
+An example of running a process: `cmdctl run ls -la`
+
+### Get process info
+The basic structure of the get-status and get-output retrieval commands will be: `get-status/get-output [<pid>]`
+
+An example of querying for process status: `cmdctl get-status 47854`
+
+We will also have commands for stopping a process and getting the process output. They will follow a similar structure as the above examples.
+
+
+### Client Authentication
+The service will utilize mTLS to authenticate and secure requests. For this implementation we will manually create a single local CA certificate that will be used to generate a client side certificate and key pair. The CA will be trusted by the server to allow client certificates to connect. These certificates and keys will be stored locally for the client.
+
+The client will enforce TLS version 1.3 for maximum security. Each client will need a certificate and key pair generated to interact with the server.
+
+When generating client certificates we will use the `-I` flag to set the user's identity. We will also use the `-n` flag to scope the client's access to their corresponding user on the host.
```

Comment:
It's not very clear to me what CLI these flags are for - I would guess perhaps openssl but I don't think I recognise these flags.

Reply:
Ah yes sorry these are for openssl.  I can remove those specific flags so the design is less tool specific.